dataRetriever/requestHandlers/requestHandler.go

```diff
@@ -278,73 +215,131 @@ func (rrh *resolverRequestHandler) RequestHeader(destShardID uint32, hash []byte
+	key := []byte(fmt.Sprintf("%d-%d", shardId, nonce))
+	if !rrh.testIfRequestIsNeeded(key) {
 		return
 	}
 
-	var err error
-	var resolver dataRetriever.Resolver
-	var topic string
-	if rrh.isMetaChain {
-		topic = rrh.shardHdrRequestTopic
-		resolver, err = rrh.resolversFinder.CrossShardResolver(topic, destShardID)
-	} else {
-		topic = rrh.metaHdrRequestTopic
-		resolver, err = rrh.resolversFinder.MetaChainResolver(topic)
+	headerResolver, err := rrh.getShardHeaderResolver(shardId)
+	if err != nil {
+		log.Error("RequestShardHeaderByNonce",
+			"error", err.Error(),
+		)
+		return
 	}
 
+	err = headerResolver.RequestDataFromNonce(nonce)
 	if err != nil {
-		log.Debug("missing resolver",
-			"topic", topic,
-			"shard", destShardID,
-		)
+		log.Debug("RequestShardHeaderByNonce", "error", err.Error())
 		return
 	}
 
-	headerResolver, ok := resolver.(dataRetriever.HeaderResolver)
-	if !ok {
-		log.Debug("resolver is not a header resolver",
-			"topic", topic,
-			"shard", destShardID,
+	rrh.addRequestedItem(key)
+}
+
+// RequestMetaHeaderByNonce method asks for meta header from the connected peers by nonce
+func (rrh *resolverRequestHandler) RequestMetaHeaderByNonce(nonce uint64) {
+	key := []byte(fmt.Sprintf("%d-%d", sharding.MetachainShardId, nonce))
+	if !rrh.testIfRequestIsNeeded(key) {
+		return
+	}
+
+	headerResolver, err := rrh.getMetaHeaderResolver()
+	if err != nil {
+		log.Error("RequestMetaHeaderByNonce",
+			"error", err.Error(),
 		)
 		return
 	}
 
 	err = headerResolver.RequestDataFromNonce(nonce)
 	if err != nil {
-		log.Debug("RequestDataFromNonce", "error", err.Error())
+		log.Debug("RequestMetaHeaderByNonce", "error", err.Error())
 		return
 	}
 
-	err = rrh.requestedItemsHandler.Add(key)
+	rrh.addRequestedItem(key)
+}
+
+func (rrh *resolverRequestHandler) testIfRequestIsNeeded(key []byte) bool {
+	rrh.sweepIfNeeded()
+
+	if rrh.requestedItemsHandler.Has(string(key)) {
+		log.Trace("item already requested",
+			"key", key)
+		return false
+	}
+
+	return true
+}
+
+func (rrh *resolverRequestHandler) addRequestedItem(key []byte) {
+	err := rrh.requestedItemsHandler.Add(string(key))
 	if err != nil {
-		log.Trace("add requested item with error",
+		log.Debug("add requested item with error",
 			"error", err.Error(),
 			"key", key)
 	}
 }
 
+func (rrh *resolverRequestHandler) getShardHeaderResolver(shardId uint32) (dataRetriever.HeaderResolver, error) {
+	isMetachainNode := rrh.shardID == sharding.MetachainShardId
+	shardIdMissmatch := rrh.shardID != shardId
+	isRequestInvalid := !isMetachainNode && shardIdMissmatch
+	if isRequestInvalid {
+		return nil, dataRetriever.ErrBadRequest
+	}
+
+	//requests should be done on the topic shardBlocks_0_META so that is why we need to figure out
+	//the cross shard id
+	crossShardId := sharding.MetachainShardId
+	if isMetachainNode {
+		crossShardId = shardId
+	}
+
+	resolver, err := rrh.resolversFinder.CrossShardResolver(factory.ShardBlocksTopic, crossShardId)
```

Comment:
This should not be the case as there is a check (a few lines before) that will make the switch, not allowing rrh.shardID and shardId to have the same sharding.MetachainShardId value. Suggesion to make this code clearer?

Reply:
No, if you check the code this case is not covered

Reply:
So, just to double check that we are both watching the same piece of code:
```GO
isMetachainNode := rrh.shardID == sharding.MetachainShardId
	shardIdMissmatch := rrh.shardID != shardId
	isRequestInvalid := !isMetachainNode && shardIdMissmatch
	if isRequestInvalid {
		return nil, dataRetriever.ErrBadRequest
	}

	//requests should be done on the topic shardBlocks_0_META so that is why we need to figure out
	//the cross shard id
	crossShardId := sharding.MetachainShardId
	if isMetachainNode {
		crossShardId = shardId
	}

	resolver, err := rrh.resolversFinder.CrossShardResolver(factory.ShardBlocksTopic, crossShardId)
	if err != nil {
		err = fmt.Errorf("%w, topic: %s, current shard ID: %d, cross shard ID: %d",
			err, factory.ShardBlocksTopic, rrh.shardID, crossShardId)
		return nil, err
	}
```

I was talking about this if section
```GO
crossShardId := sharding.MetachainShardId
if isMetachainNode {
		crossShardId = shardId
	}
```

so, crossShardId is either `sharding.MetachainShardId` or `shardId` . The condition is put in such manner that, if this was called from a meta node, it will yield `shardId` otherwise  `sharding.MetachainShardId`. Am I missing something?

Reply:
Exactly about this case I was talking about. When this method is called from a meta node with shardID parameter -> sharding.MetachainShardId. In this case the resulted topic in line 305 would be shardBlocks_META. And I was asking if this is correct or not?

Reply:
In this case crossShardId would be sharding.MetachainShardId even if it would be set in line 300 or 302

Reply:
Got it (finally). Fixed so it will output an error if the method is called with the parameter holding the sharding.MetachainID value